### PR TITLE
Support for Tree Search Policy Operator(s)

### PR DESCRIPTION
**Pull request by wenkesj**
First of all, I know this is coming out of the _blue_:

I do a lot of scrappy RL work on the side and don't really see concrete foot-work on the movement toward useful RL frameworks, so I see this as kind of a breath-of-fresh-air that you are working in Python + Tensorflow now ;). I do see other movements like [TensorForce](https://github.com/reinforceio/TensorForce/) and [Ray](https://github.com/ray-project/ray), but still nothing worth my spare-time as a dude "who just experiments" (for the time being).

# What's in the box

+ A naive, abstract tree-like policy operator and a **Monte Carlo Tree Search** implementation for **Gym** environments and **any-rl** models.  The method I used is written in Python, without any parallelization or unique optimizations (yet), and is loosely-inspired by a few methods I've seen out there using the **UCT** algorithm `Search = (Act -> Expand -> Backup)`. 

+ This attempts to maximize the use of Multi-/Discrete/Binary action space environments in **Gym** without changing the API for Rollouts. The ultimate goal would be to demonstrate, as a possible use-case, the _AlphaGo Zero_ algorithm, among others. Although, "multi-player" **Gym** environments may require _perspective_ (`*{-1, 1}`) work in the reward/observation/action spaces.

+ This example provided demonstrates the use of the MCTS model operator on a MLP using PPO to train on the returned MCTS "rollout". The `Roller` is implemented as a `TreeRoller` and can be used to generate easily MCTS guided rollouts and train policies.

_Aside_, I also added a .gitignore file, which I can also get rid of, but it's very OS friendly for developers.

# Conclusions

There may be bugs, since I did this fairly quick, but it is interesting to train, although fairly slow since the MCTS is in pure python and not parallelized.

Slow, gradual improvement, but the hyperparameters need work:
```
batch 0: mean=12.233333
...
batch 5: mean=16.766667
...
batch 10: mean=19.033333
...
batch 15: mean=22.133333
...
batch 20: mean=30.933333
```

### CODE REVIEW

**Comment by unixpickle** [comment]
Won't get time to look at this for a few more days. Hang tight!

**Comment by wenkesj** [comment]
@unixpickle no problemo, this is [et. could be] largely a work-in-progress and exploratory, so hang tight is right!

**Comment by unixpickle** [comment]
After looking through this a bit, I'd say it's a little too experimental to merge. It would probably take a lot more planning/design to get something stable/fast/general enough for me to put it into a release of anyrl-py. However, perhaps it'd be worth having an `experimental` sub-module for things like this, where breaking changes are common and expected.

Right now, I don't have very much free time for side projects. As a result, I probably won't be able to collaborate on this, at least not in the near future.

While I was skimming the code, I did notice a few superficial things:
 * AFAICT, you use `copy.deepcopy(env)` to copy environment states. I don't expect this to work for environments that are implemented internally by C objects. In order to fix this, we'd probably want to extend the `Env` abstraction to implement state saving/restoration. This would be something to think a lot about during the design stages.
 * I wrap all doc comments after 60 characters.

**Comment by wenkesj** [comment]
@unixpickle I agree with all the points, obviously it would take a lot of time to make right. This was kind of a spur of the moment thing, and I don't have much free time either. Maybe anytree will see the light here some day. Keep up the good work.
